{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "import pickle\n",
    "from collections import defaultdict\n",
    "from scipy import stats\n",
    "import matplotlib.pyplot as plt\n",
    "from collections import Counter\n",
    "import networkx as nx\n",
    "from multiomics_benchmark import helper_functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define your datasets\n",
    "datasets = ['brain','breast', 'colorectal', 'glioma', 'head_neck', 'kidney_renal','lung', \n",
    "            'lung_squamous',  'omics', 'ovarian', 'pan_kidney', 'PreE', \n",
    "            'stomach_esophageal', 'thyroid', 'uterine']\n",
    "\n",
    "# Define omics names for each dataset\n",
    "options = defaultdict(lambda: [\"Methylation\", \"miRNA\", \"RNA\", \"RPPA\", \"SCNV\"], \n",
    "                      {'omics': ['cfRNA', 'proteo_lum', 'serum_lum', 'microbiome', 'CyTOF', 'metabolome', 'proteo_som'],\n",
    "                       'PreE': ['cfRNA', 'lipidome', 'metabol-plasma', 'metabol-urine', 'proteome', 'microbiome'],\n",
    "                       'glioma': [\"Methylation\", \"miRNA\", \"RNA\", \"RPPA\", \"SCNV\"],\n",
    "                       'brain': [\"Methylation\", \"miRNA\", \"RNA\", \"RPPA\", \"SCNV\"],\n",
    "                       'ovarian': [\"Methylation\", \"miRNA\", \"RNA\", \"RPPA\", \"SCNV\"],\n",
    "                       'uterine':[\"Methylation\", \"miRNA\", \"RPPA\", \"SCNV\"]\n",
    "                      })"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calculate accuracy for optimised and default models "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "#What you want your accuracy threshold to be\n",
    "r_limit = 0.5\n",
    "r_limit_name = 'limit_{}'.format(r_limit)\n",
    "\n",
    "r_p_all = []\n",
    "n_samples = 50"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'stomach_esophageal'"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Run this to get results for default models \n",
    "for dataset in datasets:\n",
    "    \n",
    "    omics = options[dataset]\n",
    "\n",
    "    os.chdir('/home/mxenoc/workspace/benchmark/results/'+dataset+'/off_the_shelf/')\n",
    "\n",
    "    feats_list = []\n",
    "\n",
    "    for i in range(len(omics)): feats_list.append(np.tile(omics[i],n_samples))\n",
    "        \n",
    "    feats = np.hstack(feats_list)\n",
    "\n",
    "    EN = dict()\n",
    "    GFA = dict()\n",
    "    Lasso = dict()\n",
    "    MT_Lasso = dict()\n",
    "    RF = dict()\n",
    "    SGLasso = dict()\n",
    "    blockForest = dict()\n",
    "    Ridge = dict()\n",
    "\n",
    "    mdl =      [EN,    Lasso,   SGLasso,   RF,    Ridge,   GFA,   MT_Lasso]\n",
    "    mdl_name = ['EN', 'Lasso', 'SGLasso', 'RF',  'Ridge', 'GFA', 'MT_Lasso']\n",
    "\n",
    "    for k in range(len(mdl_name)):\n",
    "        with open(f\"{mdl_name[k]}.pkl\", 'rb') as f:\n",
    "            mdl[k] = pickle.load(f)\n",
    "\n",
    "    r_p_list = []\n",
    "    models_res = []\n",
    "    res = []    \n",
    "\n",
    "    for i in range(len(mdl_name)-2): #loop through the models\n",
    "\n",
    "        all_features = defaultdict(list)\n",
    "\n",
    "        for l in range(len(omics)):   #loop through the modalities\n",
    "            each_feature = defaultdict(list)\n",
    "\n",
    "            for k in range(n_samples):   #loop through the targets\n",
    "\n",
    "                x = mdl[i]['observed_test'][l][k]\n",
    "                y = mdl[i]['prediction_test'][l][k]\n",
    "                y.iloc[0,0] += 0.0000000001\n",
    "\n",
    "                each_feature['r'].append(stats.spearmanr(x, y)[0])                \n",
    "                each_feature['p'].append(stats.spearmanr(x, y)[1])\n",
    "\n",
    "            all_features['r'].append(each_feature['r'])\n",
    "            all_features['p'].append(each_feature['p'])\n",
    "\n",
    "        r = np.hstack(all_features['r'])\n",
    "        p = np.hstack(all_features['p'])\n",
    "\n",
    "        r_p = pd.DataFrame({'r': r, 'p': p, 'omics': feats, 'model': mdl_name[i], 'data' : dataset})\n",
    "        r_p = r_p[(r_p['r']>r_limit) & (r_p['p']<=0.05)]\n",
    "        r_p_list.append(r_p)\n",
    "        res.append(len(r_p))\n",
    "\n",
    "    for n_GFA in ([len(mdl_name)-2, len(mdl_name)-1]):\n",
    "\n",
    "        all_features = defaultdict(list)\n",
    "\n",
    "        for l in range(len(omics)):\n",
    "            each_feature = defaultdict(list)\n",
    "\n",
    "            for k in range(n_samples):\n",
    "\n",
    "                x = mdl[n_GFA]['observed_test'][l][k]\n",
    "                y = mdl[n_GFA]['prediction_test'][l].iloc[:,k]\n",
    "                                \n",
    "                each_feature['r'].append(stats.spearmanr(x, y)[0])                \n",
    "                each_feature['p'].append(stats.spearmanr(x, y)[1])\n",
    "\n",
    "            all_features['r'].append(each_feature['r'])\n",
    "            all_features['p'].append(each_feature['p'])\n",
    "\n",
    "        r = np.hstack(all_features['r'])\n",
    "        p = np.hstack(all_features['p'])\n",
    "\n",
    "        r_p = pd.DataFrame({'r': r, 'p': p, 'omics': feats, 'model': mdl_name[n_GFA], 'data' : dataset})\n",
    "        r_p = r_p[(r_p['r']>r_limit) & (r_p['p']<=0.05)]\n",
    "        r_p_list.append(r_p)\n",
    "        res.append(len(r_p))\n",
    "\n",
    "    results = pd.DataFrame({'features': np.hstack(res), 'models': np.hstack(mdl_name), 'dataset': dataset})\n",
    "    results['percentage'] = results['features']/(len(omics)*n_samples)*100\n",
    "\n",
    "    r_p_all.append(pd.concat(r_p_list))\n",
    "    \n",
    "    os.chdir('/home/mxenoc/workspace/pickles/multiomics/plots/accuracy_calculations/default/'+r_limit_name)\n",
    "    with open(dataset+'.pkl', 'wb') as f:  \n",
    "        pickle.dump(results, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Run this to get results for optimised models \n",
    "for dataset in datasets:\n",
    "    \n",
    "    omics = options[dataset]\n",
    "\n",
    "    os.chdir('/home/mxenoc/workspace/benchmark/results/'+dataset+'/optimised/')\n",
    "\n",
    "    feats_list = []\n",
    "\n",
    "    for i in range(len(omics)): feats_list.append(np.tile(omics[i],n_samples))\n",
    "        \n",
    "    feats = np.hstack(feats_list)\n",
    "\n",
    "    EN = dict()\n",
    "    GFA = dict()\n",
    "    Lasso = dict()\n",
    "    MT_Lasso = dict()\n",
    "    RF = dict()\n",
    "    group_lasso = dict()\n",
    "    blockForest = dict()\n",
    "    ridge = dict()\n",
    "\n",
    "    mdl =      [EN,    Lasso,   group_lasso,    ridge,   blockForest,   GFA,   MT_Lasso]\n",
    "    mdl_name = ['EN', 'Lasso', 'group_lasso',  'ridge', 'blockForest', 'GFA', 'MT_Lasso']\n",
    "\n",
    "    for k in range(len(mdl_name)):\n",
    "        with open(f\"{mdl_name[k]}.pkl\", 'rb') as f:\n",
    "            mdl[k] = pickle.load(f)\n",
    "\n",
    "    r_p_list = []\n",
    "    models_res = []\n",
    "    res = []    \n",
    "\n",
    "    for i in range(len(mdl_name)-2): #loop through the models\n",
    "\n",
    "        all_features = defaultdict(list)\n",
    "\n",
    "        for l in range(len(omics)):   #loop through the modalities\n",
    "            each_feature = defaultdict(list)\n",
    "\n",
    "            for k in range(n_samples):   #loop through the targets\n",
    "\n",
    "                x = mdl[i]['observed_test'][l][k]\n",
    "                y = mdl[i]['prediction_test'][l][k]\n",
    "                y.iloc[0,0] += 0.0000000001\n",
    "\n",
    "                each_feature['r'].append(stats.spearmanr(x, y)[0])                \n",
    "                each_feature['p'].append(stats.spearmanr(x, y)[1])\n",
    "\n",
    "            all_features['r'].append(each_feature['r'])\n",
    "            all_features['p'].append(each_feature['p'])\n",
    "\n",
    "        r = np.hstack(all_features['r'])\n",
    "        p = np.hstack(all_features['p'])\n",
    "\n",
    "        r_p = pd.DataFrame({'r': r, 'p': p, 'omics': feats, 'model': mdl_name[i], 'data' : dataset})\n",
    "        r_p = r_p[(r_p['r']>r_limit) & (r_p['p']<=0.05)]\n",
    "        r_p_list.append(r_p)\n",
    "        res.append(len(r_p))\n",
    "\n",
    "    for n_GFA in ([len(mdl_name)-2, len(mdl_name)-1]):\n",
    "\n",
    "        all_features = defaultdict(list)\n",
    "\n",
    "        for l in range(len(omics)):\n",
    "            each_feature = defaultdict(list)\n",
    "\n",
    "            for k in range(n_samples):\n",
    "\n",
    "                x = mdl[n_GFA]['observed_test'][l][k]\n",
    "                y = mdl[n_GFA]['prediction_test'][l].iloc[:,k]\n",
    "                                \n",
    "                each_feature['r'].append(stats.spearmanr(x, y)[0])                \n",
    "                each_feature['p'].append(stats.spearmanr(x, y)[1])\n",
    "\n",
    "            all_features['r'].append(each_feature['r'])\n",
    "            all_features['p'].append(each_feature['p'])\n",
    "\n",
    "        r = np.hstack(all_features['r'])\n",
    "        p = np.hstack(all_features['p'])\n",
    "\n",
    "        r_p = pd.DataFrame({'r': r, 'p': p, 'omics': feats, 'model': mdl_name[n_GFA], 'data' : dataset})\n",
    "        r_p = r_p[(r_p['r']>r_limit) & (r_p['p']<=0.05)]\n",
    "        r_p_list.append(r_p)\n",
    "        res.append(len(r_p))\n",
    "\n",
    "    results = pd.DataFrame({'features': np.hstack(res), 'models': np.hstack(mdl_name), 'dataset': dataset})\n",
    "    results['percentage'] = results['features']/(len(omics)*n_samples)*100\n",
    "\n",
    "    r_p_all.append(pd.concat(r_p_list))\n",
    "    \n",
    "    os.chdir('/home/mxenoc/workspace/pickles/multiomics/plots/accuracy_calculations/optimised/'+r_limit_name)\n",
    "    with open(dataset+'.pkl', 'wb') as f:  \n",
    "        pickle.dump(results, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calculate accuracy for individual models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "r_p_all = []\n",
    "individual_totals = {}\n",
    "individual_predictability = {}\n",
    "individual_predictors = {}\n",
    "\n",
    "r_limit = 0.5\n",
    "r_limit_name = 'limit_{}'.format(r_limit)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "#for dataset in datasets:\n",
    "for dataset in datasets:\n",
    "\n",
    "    omics = options[dataset]\n",
    "\n",
    "    with open('/home/mxenoc/workspace/benchmark/results/'+dataset+'/individual/Lasso.pkl', 'rb') as f:\n",
    "        mdl = pickle.load(f)\n",
    "\n",
    "    r_p_list = []\n",
    "    models_res = []\n",
    "    res = []    \n",
    "\n",
    "    for omic_dataset in range(len(omics)):\n",
    "        all_features = defaultdict(list)\n",
    "        for k in range(n_samples):\n",
    "            each_feature = defaultdict(list)\n",
    "            for l in range(len(omics)-1):                \n",
    "\n",
    "                x = mdl['observed_test'][omic_dataset][k]\n",
    "                y = mdl['prediction_test'][omic_dataset][k][l]\n",
    "\n",
    "                each_feature['r'].append(stats.spearmanr(x,y)[0])                \n",
    "                each_feature['p'].append(stats.spearmanr(x,y)[1])\n",
    "                each_feature['predictor'].append(omics[l])\n",
    "\n",
    "            all_features['r'].append(each_feature['r'])\n",
    "            all_features['p'].append(each_feature['p'])\n",
    "            all_features['predictor'].append(each_feature['predictor'])\n",
    "\n",
    "        r = np.hstack(all_features['r'])\n",
    "        p = np.hstack(all_features['p'])\n",
    "        predictors = np.hstack(all_features['predictor'])\n",
    "\n",
    "        r_p = pd.DataFrame({'r': r, 'p': p, 'predictor': predictors, 'target': omics[omic_dataset]})\n",
    "        r_p = r_p[(r_p['r']>r_limit) & (r_p['p']<=0.05)]\n",
    "        r_p_list.append(r_p)\n",
    "        res.append(len(r_p))\n",
    "\n",
    "    test = pd.concat(r_p_list)\n",
    "    test = test.drop(['r', 'p'], axis = 1)\n",
    "\n",
    "    dataset_predictability = Counter(test['target'])\n",
    "    dataset_predictability = pd.DataFrame(dataset_predictability.items())\n",
    "    dataset_predictability.columns = ['features', 'group']\n",
    "\n",
    "    dataset_predictors = Counter(test['predictor'])\n",
    "    dataset_predictors = pd.DataFrame(dataset_predictors.items())\n",
    "    dataset_predictors.columns = ['features', 'group']\n",
    "\n",
    "    links = pd.DataFrame(test.groupby(test.columns.tolist(),as_index=False).size())\n",
    "    links.reset_index(level=0, inplace=True)\n",
    "    links.reset_index(level=0, inplace=True)\n",
    "    columns_titles = [\"predictor\",\"target\", 0]\n",
    "    links = links.reindex(columns = columns_titles)\n",
    "    links.columns = ['var1', 'var2', 'weight']\n",
    "\n",
    "    ###DO NOT USE THIS FOR MULTIOMICS AND PREE\n",
    "    links['var1'] = links['var1'].str.partition('_')[0]\n",
    "    links['var2'] = links['var2'].str.partition('_')[0]\n",
    "    #######\n",
    "\n",
    "    G = nx.from_pandas_edgelist(links, 'var1', 'var2', edge_attr=True, create_using=nx.DiGraph())\n",
    "\n",
    "    # Edges you want to include\n",
    "    threshold = 0\n",
    "\n",
    "    # Edges you want to plot\n",
    "    edges_filtered = [(u,v) for (u, v, e) in G.edges(data=True) if e['weight'] > threshold]\n",
    "    weights_filtered = [e['weight'] for (u, v, e) in G.edges(data=True) if e['weight'] > threshold]\n",
    "\n",
    "    # Choose the layout\n",
    "    pos = nx.spring_layout(G, scale = 0.5, k = 0.65)\n",
    "\n",
    "    node_colors = dataset_predictability.set_index('features')\n",
    "    # Reindex your nodes to match the graph's nodes\n",
    "    node_colors_reind = node_colors.reindex(G.nodes())\n",
    "    #node_colors_reind['group'] = pd.Categorical(node_colors_reind['group'])\n",
    "    node_colors_percent = (node_colors_reind/((len(omics)-1)*50))*100\n",
    "\n",
    "    popular_predictors = dataset_predictors.set_index('features')\n",
    "    # Reindex your nodes to match the graph's nodes\n",
    "    popular_predictors_reind = popular_predictors.reindex(G.nodes())\n",
    "    #node_colors_reind['group'] = pd.Categorical(node_colors_reind['group'])\n",
    "    popular_predictors_percent = (popular_predictors_reind/((len(omics)-1)*50))*100\n",
    "\n",
    "    individual_totals[dataset] = popular_predictors_percent\n",
    "    individual_predictors[dataset] = dataset_predictors\n",
    "    individual_predictability[dataset] = dataset_predictability"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'brain':       features  group\n",
       " 0        miRNA     91\n",
       " 1  Methylation     89\n",
       " 2          RNA     87\n",
       " 3         RPPA     23, 'breast':       features  group\n",
       " 0  Methylation    110\n",
       " 1        miRNA     73\n",
       " 2          RNA     84\n",
       " 3         RPPA     23, 'colorectal':       features  group\n",
       " 0        miRNA     61\n",
       " 1  Methylation     75\n",
       " 2          RNA     56\n",
       " 3         RPPA     17, 'glioma':       features  group\n",
       " 0  Methylation     97\n",
       " 1        miRNA    108\n",
       " 2          RNA     92\n",
       " 3         RPPA     28, 'head_neck':       features  group\n",
       " 0        miRNA     67\n",
       " 1         RPPA     25\n",
       " 2  Methylation     83\n",
       " 3          RNA     76, 'kidney_renal':       features  group\n",
       " 0  Methylation     38\n",
       " 1        miRNA     40\n",
       " 2          RNA     38\n",
       " 3         RPPA      9, 'lung':       features  group\n",
       " 0        miRNA     57\n",
       " 1  Methylation     63\n",
       " 2          RNA     67\n",
       " 3         RPPA     13, 'lung_squamous':       features  group\n",
       " 0        miRNA     59\n",
       " 1         RPPA     32\n",
       " 2  Methylation     88\n",
       " 3          RNA     66, 'omics':      features  group\n",
       " 0  proteo_lum     15\n",
       " 1  metabolome      8\n",
       " 2       CyTOF      7\n",
       " 3  microbiome      3\n",
       " 4   serum_lum      1, 'ovarian':       features  group\n",
       " 0        miRNA     51\n",
       " 1          RNA     81\n",
       " 2  Methylation     66\n",
       " 3         RPPA     30, 'pan_kidney':       features  group\n",
       " 0  Methylation    106\n",
       " 1        miRNA     91\n",
       " 2         RPPA     41\n",
       " 3          RNA     68, 'PreE':          features  group\n",
       " 0  metabol-plasma     11\n",
       " 1           cfRNA      8, 'stomach_esophageal':       features  group\n",
       " 0        miRNA     72\n",
       " 1  Methylation     92\n",
       " 2          RNA     69\n",
       " 3         RPPA      5, 'thyroid':       features  group\n",
       " 0  Methylation     54\n",
       " 1        miRNA     65\n",
       " 2          RNA     50, 'uterine':       features  group\n",
       " 0  Methylation     49\n",
       " 1        miRNA     14\n",
       " 2         RPPA      2}"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "individual_predictors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/home/mxenoc/workspace/pickles/multiomics/plots/individual_totals.pkl', 'wb') as f:  \n",
    "    pickle.dump(individual_totals, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calculate accuracy for stacked models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create an empty list for the results\n",
    "results_all = list()\n",
    "n_samples = 50"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "metalearner_list = list()\n",
    "metalearner_list.append(['opt', 'Lasso', False])\n",
    "metalearner_list.append(['opt', 'Lasso', True])\n",
    "metalearner_list.append(['opt', 'EN', False])\n",
    "metalearner_list.append(['opt', 'EN', True])\n",
    "metalearner_list.append(['regular', 'RF', False])\n",
    "metalearner_list.append(['opt', 'Ridge', False])\n",
    "metalearner_list.append(['nnls', 'nnls', False])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "metalearner_list = list()\n",
    "metalearner_list.append(['regular', 'lasso', False])\n",
    "metalearner_list.append(['regular', 'lasso', True])\n",
    "metalearner_list.append(['regular', 'EN', False])\n",
    "metalearner_list.append(['regular', 'EN', True])\n",
    "metalearner_list.append(['regular', 'RF', False])\n",
    "metalearner_list.append(['regular', 'ridge', False])\n",
    "metalearner_list.append(['nnls', 'nnls', False])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "  0%|          | 0/15 [00:00<?, ?it/s]\u001b[A\u001b[A\n",
      "\n",
      "  7%|â–‹         | 1/15 [00:13<03:08, 13.48s/it]\u001b[A\u001b[A"
     ]
    },
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'Stacked_Lasso.pkl'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-91-0a15a9adfea7>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     19\u001b[0m         \u001b[0mmetalearner_names\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmetalearner_list\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m'_'\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mmetalearner_list\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m'_'\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmetalearner_list\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     20\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 21\u001b[0;31m     \u001b[0;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Stacked_Lasso.pkl\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'rb'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     22\u001b[0m         \u001b[0mmdl\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpickle\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mload\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     23\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'Stacked_Lasso.pkl'"
     ]
    }
   ],
   "source": [
    "r_p_all = []\n",
    "\n",
    "for dataset_index in tqdm(range(len(datasets))):\n",
    "\n",
    "    dataset = datasets[dataset_index]\n",
    "    omics = options[dataset]\n",
    "\n",
    "    os.chdir('/home/mxenoc/workspace/benchmark/results/'+dataset+'/optimised')\n",
    "\n",
    "    feats_list = []\n",
    "\n",
    "    for i in range(len(omics)): \n",
    "        feats_list.append(np.tile(omics[i],n_samples))\n",
    "\n",
    "    feats = np.hstack(feats_list)\n",
    "\n",
    "    metalearner_names= list()\n",
    "    for k in range(len(metalearner_list)):\n",
    "        metalearner_names.append(metalearner_list[k][0]+'_'+metalearner_list[k][1]+'_'+str(metalearner_list[k][2]))\n",
    "\n",
    "    with open(\"Stacked_Lasso.pkl\", 'rb') as f:\n",
    "        mdl = pickle.load(f)\n",
    "\n",
    "    r_p_list = []\n",
    "    #res = defaultdict(list)\n",
    "    models_res = []\n",
    "\n",
    "    res = []    \n",
    "    for metalearners in range(len(metalearner_names)):\n",
    "        all_features = defaultdict(list)\n",
    "        for l in range(len(omics)):\n",
    "            each_feature = defaultdict(list)\n",
    "\n",
    "            for k in range(n_samples):\n",
    "\n",
    "                x = mdl['observed_test'][l][k]\n",
    "                y = pd.concat(mdl['prediction_test'][l][k][metalearner_names[metalearners]])\n",
    "                y.iloc[0,0] += 0.0000000001\n",
    "\n",
    "                each_feature['r'].append(stats.spearmanr(x,y)[0])                \n",
    "                each_feature['p'].append(stats.spearmanr(x,y)[1])\n",
    "\n",
    "            all_features['r'].append(each_feature['r'])\n",
    "            all_features['p'].append(each_feature['p'])\n",
    "\n",
    "        r = np.hstack(all_features['r'])\n",
    "        p = np.hstack(all_features['p'])\n",
    "\n",
    "        r_p = pd.DataFrame({'r': r, 'p': p, 'omics': feats, 'model': metalearner_names[metalearners], 'data' : dataset})\n",
    "        r_p = r_p[(r_p['r']>r_limit) & (r_p['p']<=0.05)]\n",
    "        r_p_list.append(r_p)\n",
    "        res.append(len(r_p))\n",
    "\n",
    "    results = pd.DataFrame({'features': np.hstack(res), 'models': np.hstack(metalearner_names), 'dataset': dataset})\n",
    "    results['percentage'] = results['features']/(len(omics)*n_samples)*100\n",
    "\n",
    "    results_all.append(results)\n",
    "    \n",
    "    r_p_all.append(pd.concat(r_p_list))\n",
    "\n",
    "    os.chdir('/home/mxenoc/workspace/pickles/multiomics/plots/accuracy_calculations/stacked/'+r_limit_name)\n",
    "    with open(dataset+'.pkl', 'wb') as f:  \n",
    "        pickle.dump(results, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'breast'"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
